src/Components/Endpoints/src/RazorComponentEndpoint.cs

```diff
@@ -35,25 +41,94 @@ public static RequestDelegate CreateRouteDelegate(Type componentType)
                 { nameof(RazorComponentEndpointHost.ComponentParameters), componentParameters },
             });
 
+            await using var writer = CreateResponseWriter(httpContext.Response.Body);
+
             // Note that we always use Static rendering mode for the top-level output from a RazorComponentResult,
             // because you never want to serialize the invocation of RazorComponentResultHost. Instead, that host
             // component takes care of switching into your desired render mode when it produces its own output.
-            var htmlContent = await componentPrerenderer.PrerenderComponentAsync(
+            var htmlContent = (EndpointHtmlRenderer.PrerenderedComponentHtmlContent)(await endpointHtmlRenderer.PrerenderComponentAsync(
                 httpContext,
                 typeof(RazorComponentEndpointHost),
                 RenderMode.Static,
-                hostParameters);
+                hostParameters,
+                waitForQuiescence: preventStreamingRendering));
 
-            await using var writer = CreateResponseWriter(httpContext.Response.Body);
-            await htmlContent.WriteToAsync(writer);
+            // Importantly, we must not yield this thread (which holds exclusive access to the renderer sync context)
+            // in between the first call to htmlContent.WriteTo and the point where we start listening for subsequent
+            // streaming SSR batches. Otherwise some other code might dispatch to the renderer sync context and cause
+            // a batch that would get missed.
+            htmlContent.WriteTo(writer, HtmlEncoder.Default); // Don't use WriteToAsync, as per the comment above
```

Comment:
I don't think it's disallowed since `ViewBuffer` itself also uses synchronous writes whenever it has any `IHtmlContent` (also, it works).

Changing to asynchronous writes would involve much more than calling `WriteToAsync` here - we'd have to correspondingly change all the code this ends up calling into (e.g., `HtmlComponentWriter.Write` would need to become async and every bit of markup it emits would have to starting being done asynchonously).

I do understand the reasons why async writes would be better from the point of view of the server not creating such large buffers. Achieving it would be nontrivial as even a sync context doesn't innately have any API to say not to delay all inbound work (it will always start running new work if no other thread is actively still running a synchronous work item), but I do agree we could likely find a viable way to do that by adding some new API to the dispatcher.

> That said, I think it's fine if we need to do it like this for now.

👍

Reply:
Fair enough